docs/matlab/language.md

```diff
@@ -0,0 +1,743 @@
+one of several types:
+
+```yaml
+Rec: !record
+  fields:
+    intOrFloat: [int, float]
+    intOrFloatExpandedForm:
+      - int
+      - float
+    nullableIntOrFloat:
+      - null
+      - int
+      - float
+```
+
+The `null` type in the example above means that no value (`yardl.None` in Matlab) is
+also a possibility.
+
+### Generated Union Types
+
+For the Matlab codegen, we represent unions as a [tagged
+union](https://en.wikipedia.org/wiki/Tagged_union) and we generate a class for
+each union in a model. In the example above, the union `[int, float]` would have
+a class named `Int32OrFloat32`. Each union case then has a corresponding static
+constructor method in the union class.
+
+This union class can be used like this:
+
+```matlab
+function process_my_union(u)
+    if u.isInt32()
+        fprintf("%d is an int32\n", u.value);
+    elseif u.isFloat32()
+        fprintf("%f is a float\n", u.value);
+    else
+        error("Unrecognized union type");
+    end
+end
+
+process_my_union(sandbox.Int32OrFloat32.Int32(2))
+process_my_union(sandbox.Int32OrFloat32.Float32(7.9))
+```
+
+### Union Tag Names
+
+The type case name used in union classes (`Int32` and `Float32` in the example
+above) is derived from the name of the type of each case. This means that you
+will get an error when attempting to use a type in a union that is made up of
+symbols that are not valid in an identifier. For example:
+
+```yaml
+Rec: !record
+  fields:
+    floatArrayOrDoubleArray:
+      - float[] # Error! // [!code error]
+      - double[] # Error! // [!code error]
+```
+
+There are two simple solutions to this problem. The first is to give explicit
+tag names to each union case using the expanded `!union` syntax:
+
+```yaml
+Rec: !record
+  fields:
+    floatArrayOrDoubleArray: !union
+      floatArray: float[]
+      doubleArray: double[]
+```
+
+The second option is to create [aliases](#type-aliases) for the types:
+
+```yaml
+
+FloatArray: float[]
+DoubleArray: double[]
+
+Rec: !record
+  fields:
+    floatArrayOrDoubleArray:
+      - FloatArray
+      - DoubleArray
+```
+
+The first option is usually preferred, unless the type alias is going to be used
+elsewhere. In both cases, the union type will be `FloatArrayOrDoubleArray` and
+the cases `FloatArray` and `DoubleArray`.
+
+If you don't like the generated name you can give the union an
+[alias](#type-aliases):
+
+```yaml
+ArrayUnion: !union
+  floatArray: float[]
+  doubleArray: double[]
+
+Rec: !record
+  fields:
+    arrayUnion: ArrayUnion
+```
+
+Now the union class is `ArrayUnion` with case types `FloatArray` and `DoubleArray`.
+
+## Enums
+
+Enums can be defined as a list of values:
+
+```yaml
+Fruits: !enum
+  values:
+    - apple
+    - banana
+    - pear
+```
+
+Enums are generated as custom Matlab class definitions, *not* using Matlab's `enumeration` support, which doesn't allow integer values that are not explicitly defined in the enum definition.
```

Comment:
Did you consider using Matlab's enums, and retuning integers if the read value is outside the defined values?

Reply:
Yes, this is possible, however there is no good way to construct an enumeration from an integer outside the defined values. This means, for example, you can't mix an enumeration and equivalent integer in MATLAB arrays, which breaks yardl support for flags/enums in streams, vectors, and arrays:

```matlab
>> fs = [Fruits.APPLE, uint64(7)]
Error using Fruits
Cannot find a member of the 'Fruits' enumeration class that corresponds to each element of the given input
argument.
```